Correctly attribute implicit int to pointer conversion removal to C89/C90

thomasnyman opened the pull request:
- Correctly attribute implicit int to pointer conversion removal to C89/C90
- Fix incorrect URL to SEI CER C rule INT36-C

Fixes issues brought up in [comment](https://github.com/ossf/wg-best-practices-os-developers/commit/4106fb09ce1a32f118bc774309b646cc62b1256a#r139194996) to #357.

gkunz commented on docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:
```diff
@@ -366,7 +366,7 @@ Zero-warning policies can also be enforced at CI level. CI-based zero- or bounde
 Make the compiler treat obsolete C constructs as errors.
 
-The ISO/IEC 9899:1999 standard, commonly referred to as C99, removed several backwards compatibility features, such as implicit function declarations, implicit return types and certain implicit type conversion from the C language. Such implicit declarations[^DCL31-C] and type conversions[^INT36-C] can be considered dangerous for the correctness and security of C code as they lead to less stringen ...
+The ISO/IEC 9899:1999 standard, commonly referred to as C99, removed several backwards compatibility features, such as implicit function declarations and implicit return types from the C language. Similarly, the earlier C89/C90 standard (ANSI X3.159-1989 / ISO/IEC 9899:1990) removed certain implicit type conversion, such as implicit conversions from integer to pointer types. Such implicit declara ...
```
Typo:
```suggestion
The ISO/IEC 9899:1999 standard, commonly referred to as C99, removed several backwards compatibility features, such as implicit function declarations and implicit return types from the C language. Similarly, the earlier C89/C90 standard (ANSI X3.159-1989 / ISO/IEC 9899:1990) removed certain implicit type conversion, such as implicit conversions from integer to pointer types. Such implicit declarations[^DCL31-C] and type conversions (whether implicit or explicit[^INT36-C]) can be considered dangerous for the correctness and security of C code as they lead to less stringent type checking and may rely on implementation-defined behavior. Later dialects of C consider these issues as errors by default. However, modern compilers still accept these obsolete construct by default unless instructed to pedantically give errors whenever the base standard requires them.
```

david-a-wheeler replied:
I would delete the next to last sentence, "Later dialects of C consider these issues as errors by default." At this point it's more confusing than helpful, since compilers accept them.

thomasnyman commented on docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:
```diff
@@ -366,7 +366,7 @@ Zero-warning policies can also be enforced at CI level. CI-based zero- or bounde
 Make the compiler treat obsolete C constructs as errors.
 
-The ISO/IEC 9899:1999 standard, commonly referred to as C99, removed several backwards compatibility features, such as implicit function declarations, implicit return types and certain implicit type conversion from the C language. Such implicit declarations[^DCL31-C] and type conversions[^INT36-C] can be considered dangerous for the correctness and security of C code as they lead to less stringen ...
+The ISO/IEC 9899:1999 standard, commonly referred to as C99, removed several backwards compatibility features, such as implicit function declarations and implicit return types from the C language. Similarly, the earlier C89/C90 standard (ANSI X3.159-1989 / ISO/IEC 9899:1990) removed certain implicit type conversion, such as implicit conversions from integer to pointer types. Such implicit declara ...
```
Delete: "Later dialects of C consider these issues as errors by default."

thomasnyman commented:
Addressed the comments brought up during the [C/C++ Compiler BP Guide call on 2024-03-14](https://docs.google.com/document/d/1JY8FREBPCUUFpuv7-4B9EjeS2MLDpel0dbG5DFWrTns/edit#bookmark=id.jdwmn4p6l6ea).

Merging as per discussion.
